Tag pagination (#470)
opencontainers/distribution-spec: spec.md

`<name>` is the namespace of the repository, and `<int>` is an integer specifying the number of tags requested.
The response to such a request MAY return fewer than `<int>` results, but only when the total number of tags attached to the repository is less than `<int>`.
Otherwise, the response MUST include `<int>` results.

> A `Link` header MAY be included in the response when additional responses are available.

Reviewer: ```suggestion
A `Link` header MAY be included in the response when additional tags are available.
```

May be more explicit in this context.

Author: Done.

When using the `last` query parameter, the `n` parameter is OPTIONAL.

> The registry SHOULD support at least `n=100` tags.

Reviewer: ```suggestion
The registry SHOULD support at least `n=100` tags in a single response.
```

Author: Done.

> The registry SHOULD support at least `n=100` tags in a single response.

Reviewer 2: can you link to the reason 100 is the should here..

Author: It's arbitrary. I'm happy to adjust it to another value.

Reviewer: crane uses [1000 by default](https://github.com/google/go-containerregistry/blob/dbcd01c402b2f05bcf6fb988014c5f37e9b13559/pkg/v1/remote/options.go#L108) which apparently works with all current registries so maybe we can make this n=1000.

Author: It looks like quay.io actually caps out at 50, I thought they were more like 250, so I may have picked a number that's too large for current registries. 

```
curl -sSL -H 'Accept: application/json' https://quay.io/v2/openshift-release-dev/ocp-release/tags/list?n=1000  | jq .tags
```

Reviewer 3: Is it really worth specifying a minimum?

Author: The "SHOULD" means it's advice that well behaved registries would follow, but there's no enforcement. The goal is to get some guidance that most implementations agree with so that servers and clients aren't both picking arbitrary numbers. Increasing this to 1000 if no additional major registries are excluded feels reasonable.

Author: I removed the recommended limit from the PR.

> The client SHOULD avoid requesting more than `n=100` tags.

Reviewer: ```suggestion
The client SHOULD avoid requesting more than `n=100` tags in a single request.
```

Author: Done.

> If the registry is unable to return all tags requested by the client, the registry MUST return a `400 Bad Request` with `TAG_LIMIT_EXCEEDED` error <sup>[code-15](#error-codes)</sup>.

Reviewer: Maybe it would make sense to require the maximum number of tags to be included in the response in this case so that clients don't have to guess the maximum number (and potentially issue more unservable requests).

Author: Makes sense. The goal with the `n=100` was to give clients a number they know should be a safe default. Do you believe registries will try to go lower than that? Or is this planning for clients that try to pull 1000 tags and the registry bumps them down to 250?

Should this be done with an OCI header in the error response?

Reviewer: Honestly I wouldn't know why registries should go lower than 100 as tag lookup shouldn't be a costly operation, anyway. But on the client side I would expect clients to try to go over that to compile a list of all tags (e.g. for filtering/searching purposes). In that case it would incur less complexity on the client side (implement an algorithm for finding the lower barrier) when the registry would just say "nope, 1000 is too much, but 250 is good" and the client would only issue 2 requests in total to get to that.

About how to return that limit I'm unsure as the "Error Codes" section is very open with regards to what registries should return, e.g. registries may choose to not return a body at all so a mandate to set an OCI header in the response sounds reasonable. IIRC there is no precedence for requirements of a certain header to present in 4xx responses so this would be kind of an outlier but IMHO a well justified one.

OCI-Tag-Limit?

When using the `last` query parameter, the `n` parameter is OPTIONAL.

> The registry SHOULD support at least `n=1000` tags in a single response.

Reviewer 2: ```suggestion
Registries MAY restrict the maximum number of tags to be returned. For example, a limit of `n=1000` tags in a single response.
```

Author: I removed the recommended limit from the PR.

> The client SHOULD avoid requesting more than `n=1000` tags in a single request.

Reviewer 2: yet a very large number would be a means to identify what the limit is...

```suggestion
```

Author: A request without any limit would return the OCI-Tag-Limit header on an error when the limit is reached, making it a more efficient way to discover the limit on an as-needed basis (the registry wouldn't fail when the returned number of tags is less than its limit).

Reviewer 2: That sounds like another/different client SHOULD recommendation.

Author: It's a MUST requirement on the registry that comes out of the next line: "If the registry is unable to return all tags requested by the client, the registry MUST return a `413 Content Too Large`..."

Author: I removed the suggested limit from the PR.

> If the registry is unable to return all tags requested by the client, the registry MUST return a `413 Content Too Large` with `TAG_LIMIT_EXCEEDED` error <sup>[code-15](#error-codes)</sup> and include the header `OCI-Tag-Limit: <limit>`, where `<limit>` is the maximum tags per request supported by the registry.

Reviewer 2: ```suggestion
If the registry is unable to return all tags requested by the client, the registry MUST return a `413 Content Too Large` with `TAG_LIMIT_EXCEEDED` error <sup>[code-15](#error-codes)</sup> and include the header `OCI-Tag-Limit: <limit>`, where `<limit>` is the maximum tags supported by the registry for the request.
```

Author: I'm not following the reason for this change. Is this saying the registry would give a different limit for different requests? So if a client tried to pull 1000 tags, the registry might respond that the limit is 500. And then when the client asks for 500 tags, the registry would be permitted to return an error that says the limit is now 100 tags? Or is this saying that other kinds of requests would have different limits? Perhaps an example would help.

Reviewer 2: I'm saying I don't see a reason to force all registries to adhere to this rule that they MUST have only one maximum limit for all requests to get tags.

There could be a variance by repository {local/remote, public/private, ...} or another type of variance like total size of response.

Author: That seems like it would add complication on both clients and registries. Are registries doing this today? If not, it would be useful for clients to be able to reuse a previously seen limit from a registry to avoid additional round trips.

Reviewer 4: Adding this error will break existing correct (IMO) clients that do the right (IMO) thing by following Link headers when present so that we can loudly break any clients that are silently incorrect because they don't follow a Link header. This doesn't seem like a good tradeoff to me.

When using the `last` query parameter, the `n` parameter is OPTIONAL.

If the registry is unable to return all tags requested by the client, the registry MUST return a `413 Content Too Large` with `TAG_LIMIT_EXCEEDED` error <sup>[code-15](#error-codes)</sup> and include the header `OCI-Tag-Limit: <limit>`, where `<limit>` is the maximum tags per request supported by the registry.

*Implementers note:*
Most registries support at least 1000 tags per request, but limits have been seen as low as 50 tags per request.
To maximize the tags per page, a request without setting `n` can be used, which should either return the full list of tags, or an error with the `OCI-Tag-Limit` header indicating the limit.

> When available, clients should prefer the `Link` header over using the `last` parameter for pagination.

Reviewer 4: I would promote this from implementers note to the actual spec language somewhere near this line:

> A `Link` header MAY be included in the response when additional tags are available.

Author: I updated line 532 and reworded these notes.

| end-5   | `PATCH`        | `/v2/<name>/blobs/uploads/<reference>`                         | `202`       | `404`/`416`       |
| end-6   | `PUT`          | `/v2/<name>/blobs/uploads/<reference>?digest=<digest>`         | `201`       | `404`/`400`       |
| end-7   | `PUT`          | `/v2/<name>/manifests/<reference>`                             | `201`       | `404`             |
| end-8a  | `GET`          | `/v2/<name>/tags/list`                                         | `200`       | `404`/`400`       |

> | end-8b  | `GET`          | `/v2/<name>/tags/list?n=<integer>&last=<integer>`              | `200`       | `404`/`400`       |

Reviewer 5: Just want to note I was confused when reading this section, because in a section above it says `&last=<tagname>`, and additionally "`<tagname>` MUST NOT be a numerical index, but rather it MUST be a proper tag".

Author: Rebased to include #464.
